### PR TITLE
Adding wso2is:v6.1.0 clustering support

### PR DESCRIPTION
**Pull request by RukshanJS**
## Purpose

- [x] Adding support for a clustered deployment of WSO2 IS
- [x] Fixing issues faced when using clustering with WSO2 IS v6.1.x with the operator
- [x] Added support for different ways of applying configs
- [x] Support inline toml configs
- [x] Support ConfigMap applied to the cluster separately
- [x] Adding watchers to watch for changes to ConfigMap to trigger reconciliation
- [x] Support for deploying multiple instances of the Wso2Is CRD with clustering
- [x] Major refactoring of the source code following the best practices of Go 
- [x] Fixing the delays when applying changes to the custom resources  
- [x] Added kustomize-supported configurations for creating test MySQL instances 
- [x] Support custom fields in the CRD yaml
- [x] Develop an automation tool to generate Go structs to support deployment configs from a .csv mapping 
- [x] Added Go unit tests for the Wso2Is controller
- [x] Support the operator within a separate namespace "operators" 
- [x] Support and docs to run the operator within an AKS cluster 
- [x] Added a tool to run the scenario tests against the deployed IS instance easily

### CODE REVIEW

**Comment by RukshanJS** [comment]
@Mirage20 could you please review this PR

File: api/v1beta1/wso2is_types.go [hunk 1]
```diff
@@ -217,207 +227,205 @@ type Keystore struct {
...
+	BackupCode bool `json:"BackupCode,omitempty" toml:"BackupCode,omitempty"`
 	// +kubebuilder:default:=true
-	SMSOTPEnableByUserClaim bool `toml:"SMSOTPEnableByUserClaim,omitempty" json:"SMSOTPEnableByUserClaim,omitempty"`
+	SMSOTPEnableByUserClaim bool `json:"SMSOTPEnableByUserClaim,omitempty" toml:"SMSOTPEnableByUserClaim,omitempty"`
 	// +kubebuilder:default:=false
-	SMSOTPMandatory bool `toml:"SMSOTPMandatory,omitempty" json:"SMSOTPMandatory,omitempty"`
+	SMSOTPMandatory bool `json:"SMSOTPMandatory,omitempty" toml:"SMSOTPMandatory,omitempty"`
 	// +kubebuilder:default:=true
-	CaptureAndUpdateMobileNumber bool `toml:"CaptureAndUpdateMobileNumber,omitempty" json:"CaptureAndUpdateMobileNumber,omitempty"`
+	CaptureAndUpdateMobileNumber bool `json:"CaptureAndUpdateMobileNumber,omitempty" toml:"CaptureAndUpdateMobileNumber,omitempty"`
```
Comment (Mirage20) [inline api/v1beta1/wso2is_types.go]:
```suggestion
	CaptureAndUpdateMobileNumber bool `json:"captureAndUpdateMobileNumber,omitempty" toml:"captureAndUpdateMobileNumber,omitempty"`
```

The json key should be in camelCase right?

File: artifacts/04-crd-iam.wso2.com_userstores.yaml [hunk 2]
```diff
@@ -84,16 +83,14 @@ spec:
...
-        status: { }
+        status: {}
 status:
   acceptedNames:
     kind: ""
     plural: ""
-  conditions: [ ]
-  storedVersions: [ ]
+  conditions: []
+  storedVersions: []
```
Comment (Mirage20) [inline artifacts/04-crd-iam.wso2.com_userstores.yaml]: Let's add a missing newline at the end of all the files

File: artifacts/06-controller.yaml [hunk 3]
```diff
@@ -11,11 +11,11 @@ spec:
...
+        sidecar.istio.io/inject: 'false'
       labels:
         app: controller
     spec:
       containers:
         - name: controller
-          image: sureshmichael/wso2-iam-operator:latest
-      serviceAccountName: controller
\ No newline at end of file
+          image: sureshmichael/wso2-iam-operator:latest # TODO: change this
```
Comment (Mirage20) [inline artifacts/06-controller.yaml:20]: We need to update this right?

File: controllers/wso2is/cm.go [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+	volumeIndex := getVolumeIndex(sfs.Spec.Template.Spec.Volumes, instance.Name+"-config")
+	if volumeIndex != -1 {
+		// Change the ConfigMap which acts as the source for the Volume with name `variables.ConfigMap` to the default ConfigMap by getting inline configs.
+		if sfs.Spec.Template.Spec.Volumes[volumeIndex].VolumeSource.ConfigMap.LocalObjectReference.Name != configMapRefName {
+			sfs.Spec.Template.Spec.Volumes[volumeIndex].VolumeSource.ConfigMap.LocalObjectReference.Name = configMapRefName
+		}
+		currentHash, err := calculateConfigMapHash(configMap)
+
+		log.Info(sfs.Spec.Template.Annotations["configmapHash"])
+		log.Info(currentHash)
```
Comment (Mirage20) [inline controllers/wso2is/cm.go]: Do we need these logs? If we want it for troubleshooting purposes, shall we add a meaningful message to it?

File: controllers/wso2is/cm.go [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+			// Update StatefulSet with new hash of ConfigMap.
+			sfs.Spec.Template.SetAnnotations(map[string]string{
+				"configmapHash": currentHash,
+			})
+
+			if err := r.Update(ctx, sfs); err != nil {
+				return ctrl.Result{}, err
+			}
+
+			log.Info("Successfully updated the ConfigMap")
```
Comment (Mirage20) [inline controllers/wso2is/cm.go]: Does this log the config map name? If not, let's add bit more context information to it

File: controllers/wso2is/cm.go [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+
+		err = r.Get(ctx, types.NamespacedName{Name: instance.Spec.TomlConfigFile, Namespace: instance.Namespace}, configMap)
+		if err != nil {
+			log.Error(err, "Failed to get ConfigMap with the given name in CRD yaml")
+			return ctrl.Result{}, err
+		}
+		sfs := &appsv1.StatefulSet{}
+		err = r.Get(ctx, types.NamespacedName{Name: instance.Name, Namespace: instance.Namespace}, sfs)
+		if err != nil {
+			log.Info("Couldn't obtain StatefulSet. It'll be found in next reconcile loop if this is the first run.")
```
Comment (Mirage20) [inline controllers/wso2is/cm.go]:
```suggestion
			log.Error("Couldn't obtain StatefulSet. It'll be found in next reconcile loop if this is the first run.")
```

File: controllers/wso2is/ing.go [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+)
+
+func reconcileIngress(r *Wso2IsReconciler, instance wso2v1beta1.Wso2Is, log logr.Logger, err error, ctx context.Context) (ctrl.Result, error) {
+	// TODO:
+
+	// Check for ingress
+	ingressFound := networkingv1.Ingress{}
+	//logger.Info("ingress name: " + ingName + "namespace: " + instance.Namespace)
+	err = r.Get(ctx, types.NamespacedName{Name: instance.Name + "-ingress", Namespace: instance.Namespace}, &ingressFound)
+	if err != nil && errors.IsNotFound(err) {
```
Comment (Mirage20) [inline controllers/wso2is/ing.go:21]: We can simplify the condition as below,
```suggestion
	if errors.IsNotFound(err) {
```

File: controllers/wso2is/sfs.go [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+func (r *Wso2IsReconciler) defineStatefulSet(m wso2v1beta1.Wso2Is) *appsv1.StatefulSet {
+	containerImage := "rukshanjs/wso2is:v6.1.0"
```
Comment (Mirage20) [inline controllers/wso2is/sfs.go]: This needs to be a configuration

File: controllers/wso2is/role.go [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+		if errors.IsNotFound(err) {
+			log.Info("Role resource " + instance.Name + "-role" + " not found. Creating or re-creating role")
+			err = r.Create(ctx, roleDefinition)
+			if err != nil {
+				log.Error(err, "Failed to create new Role", "Role.Namespace", roleDefinition.Namespace, "Role.Name", roleDefinition.Name)
+				return ctrl.Result{}, err
+			}
+		} else {
+			log.Info("Failed to get role resource " + instance.Name + "-role" + ". Re-running reconcile.")
+			return ctrl.Result{}, err
```
Comment (Mirage20) [inline controllers/wso2is/role.go]: As we expect to rerun the loop here we need to set the `Requeue: true`

File: controllers/wso2is/role_binding.go [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+		if errors.IsNotFound(err) {
+			log.Info("RoleBinding resource " + instance.Name + "-rolebinding" + " not found. Creating or re-creating role binding")
+			err = r.Create(ctx, roleBindingDefinition)
+			if err != nil {
+				log.Error(err, "Failed to create new RoleBinding", "RoleBinding.Namespace", roleBindingDefinition.Namespace, "RoleBinding.Name", roleBindingDefinition.Name)
+				return ctrl.Result{}, err
+			}
+		} else {
+			log.Info("Failed to get roleBinding resource " + instance.Name + "-rolebinding" + ". Re-running reconcile.")
+			return ctrl.Result{}, err
```
Comment (Mirage20) [inline controllers/wso2is/role_binding.go]: Shall we check and update all the places that need to be rerun?

File: controllers/wso2is/sfs.go (hunk 8 above)
Comment (RukshanJS) [reply controllers/wso2is/sfs.go]: Would it be better to include this is a file like configs.go in the root?

File: artifacts/06-controller.yaml (hunk 3 above)
Comment (RukshanJS) [reply artifacts/06-controller.yaml:20]: This will be replaced later

File: controllers/wso2is/sfs.go (hunk 8 above)
Comment (RukshanJS) [reply controllers/wso2is/sfs.go]: Added to a /controllers/wso2is/configs.go file

**Comment by RukshanJS** [comment]
@Mirage20 I did the suggested changes. Could you please check now?

**Review by gnudeep — APPROVED** [review]

**Pull request merged**